api/v1alpha1/progressiverollout_types.go

```diff
@@ -28,18 +30,115 @@ type ProgressiveRolloutSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of ProgressiveRollout. Edit ProgressiveRollout_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// SourceRef references the resource, example an ApplicationSet, which owns ArgoCD Applications
+	//+kubebuilder:validation:Required
+	SourceRef corev1.TypedLocalObjectReference `json:"sourceRef"`
+	// Stages reference a list of Progressive Rollout stages
+	//+kubebuilder:validation:Optional
+	Stages []*ProgressiveRolloutStage `json:"stages,omitempty"`
+}
+
+// ProgressiveRolloutStage defines a rollout stage
+type ProgressiveRolloutStage struct {
+	// Name is a human friendly name for the stage
+	//+kubebuilder:validation:Required
+	Name string `json:"name"`
+	// MaxParallel is how many selected targets to update in parallel
+	//+kubebuilder:validation:Minimum:1
+	MaxParallel intstr.IntOrString `json:"maxParallel"`
+	// MaxTargets is the maximum number of selected targets to update
+	//+kubebuilder:validation:Minimum:1
+	MaxTargets intstr.IntOrString `json:"maxTargets"`
+	// Targets is the targets to update in the stage
+	//+kubebuilder:validation:Optional
+	Targets ProgressiveRolloutTargets `json:"targets,omitempty"`
+}
+
+// ProgressiveRolloutTargets defines the target of the Progressive Rollout
+type ProgressiveRolloutTargets struct {
+	// Clusters is the a cluster type of targets
+	//+kubebuilder:validation:Optional
+	Clusters Cluster `json:"clusters"`
+}
+
+// Cluster defines a target of type cluster
+type Cluster struct {
+	// Selector is a label selector to get the clusters for the update
+	//+kubebuilder:validation:Required
+	Selector metav1.LabelSelector `json:"selector"`
 }
 
 // ProgressiveRolloutStatus defines the observed state of ProgressiveRollout
 type ProgressiveRolloutStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
+	Conditions []ProgressiveRolloutCondition `json:"conditions,omitempty"`
 }
 
-// +kubebuilder:object:root=true
+// ProgressiveRolloutCondition defines a ProgressiveRollout condition
+type ProgressiveRolloutCondition struct {
+	// Type of Progressive Rollout condition
+	Type ProgressiveRolloutConditionType `json:"type"`
+	// Status of the condition, one of True, False, Unknown
+	Status metav1.ConditionStatus `json:"status"`
+	// Reason is a one-word CamelCase reason for the condition's last transition
+	Reason string `json:"reason,omitempty"`
+	// LastTransitionTime of this condition
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+}
+
+type ProgressiveRolloutConditionType string
 
+const CompletedType ProgressiveRolloutConditionType = "Completed"
```

Comment:
perhaps worth adding more types, at least "Failed"?

Reply:
I would actually name it `CompletedCondition` rather than `Type`? That's not really at type any more at this stage?

Reply:
Let's settled for `CompletedConditionType` :-)

Reply:
So, the reason is why I have only Completed, it's because that you have the `Status` that can be `true` or `false`. If we add the Failed one, than we should set Failed to true? If we look at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status they suggest to use a positive context, so we have a ProgressiveRollout that can be Completed=true or Completed=false. Does it make sense?

Reply:
I think in the DR we called the condition `RolledOut`, so if that's more clear I'm happy to change it.

Reply:
OK I get it now, thanks for the pointer.
In this case `Completed` actually makes more sense. I'm concerned `RolledOut` could potentially create confusion down the line with concepts like "Rolledback"?
From your link, we need to differentiate all those states (conditions) with the `name`.

Nice one

Reply:
As discussed, I finally get that we're going to keep it relatively light and native on the conditions, and will focus on a dedicated free form structures in `status` later to report on progress.